pallets/gear-builtin/src/lib.rs

```diff
@@ -118,8 +165,11 @@ pub trait BuiltinActor {
     /// Handles a message and returns a result and the actual gas spent.
     fn handle(
         dispatch: &StoredDispatch,
-        gas_limit: u64,
-    ) -> (Result<Payload, BuiltinActorError>, u64);
+        context: &mut BuiltinContext,
+    ) -> Result<Payload, BuiltinActorError>;
+
+    /// Returns the maximum gas that can be spent by the actor.
```

Comment:
Does  zero value get interpreted in special way?

Reply:
> Does zero value get interpreted in special way?

Not really: it'll just always pass the "initial screening" in the `BuiltinDispatcher`, but checks inside the `handle()` still remain (every time the gas counter is charged). However, if the `max_gas()` reports an accurate estimation, the `handle` would be skipped entirely and message re-queued. In the latter case we are certain no gas is going to be wasted if we still have to requeue the message after all, but only know it half way.